domain/blockdevice/state/state.go

```diff
@@ -0,0 +1,402 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/canonical/sqlair"
+	"github.com/google/uuid"
+	"github.com/juju/errors"
+
+	coredatabase "github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/blockdevice"
+)
+
+// State represents database interactions dealing with block devices.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new block device state
+// based on the input database factory method.
+func NewState(factory coredatabase.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// BlockDevices returns the BlockDevices for the specified machine.
+func (st *State) BlockDevices(ctx context.Context, machine string) ([]blockdevice.BlockDevice, error) {
```

Comment:
Do we want to handle the case where the machine doesn't exist?

Reply:
I love typed errors. A nice MachineNotFound is just asking to be added.

Reply:
We do get a not found when the machine doesn't exist already.
But I've relented and bloated this PR a little more by adding a MachineNotFound error now instead of in the next PR.